src/nbl/asset/ICPUImage.cpp

```diff
@@ -0,0 +1,319 @@
+				CBasicImageFilterCommon::clip_region_functor_t clipFunctor(subresource, range, parameters.format);
+
+				auto executePerTexelOrBlock = [&](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+				{
+					blake3_hasher_update(hasher, inData + readBlockArrayOffset, texelOrBlockByteSize);
+				};
+
+				const auto regions = image->getRegions(miplevel);
+				const bool performNullHash = regions.empty();
+
+				if (performNullHash)
+				{
+					const auto mipExtentInBlocks = info.convertTexelsToBlocks(image->getMipSize(miplevel));
+					const auto zeroLength = info.getBlockByteSize() * mipExtentInBlocks.x;
+					auto zeroArray = std::make_unique<uint8_t[]>(zeroLength);
+					for (auto z = 0; z < mipExtentInBlocks.z; z++)
+						for (auto y = 0; y < mipExtentInBlocks.y; y++)
+							blake3_hasher_update(hasher, zeroArray.get(), zeroLength);
+				}
+				else
+					CBasicImageFilterCommon::executePerRegion(std::execution::seq, image, executePerTexelOrBlock, regions, clipFunctor); // fire the hasher for a layer, note we forcing seq policy because texels/blocks cannot be handled with par policies when we hash them
+
+				blake3_hasher_finalize(hasher, reinterpret_cast<uint8_t*>(hash), sizeof(CState::hash_t)); // finalize hash for layer + put it to heap for given mip level	
+			};
+
+			std::for_each(policy, layers.begin(), layers.end(), executePerLayer); // fire per layer for given given mip level with specified execution policy, yes you can use parallel policy here if you want at it will work
+		};
+
+		std::for_each(policy, levels.begin(), levels.end(), executePerMipLevel); // fire per block of layers for given mip level with specified execution policy, yes you can use parallel policy here if you want at it will work
+
+		/*
+			scratch's heap is filled with all hashes,
+			time to use them and compute final hash
+		*/
+
+		blake3_hasher hasher;
+		blake3_hasher_init(&hasher);
+		{
+			for (auto miplevel = 0u; miplevel < parameters.mipLevels; ++miplevel)
+			{
+				const auto mipOffset = (miplevel * parameters.arrayLayers);
+
+				for (auto layer = 0u; layer < parameters.arrayLayers; ++layer)
+				{
+					auto* hash = hashes + mipOffset + layer;
+					blake3_hasher_update(&hasher, hash->data, sizeof(CState::hash_t));
+				}
+			}
+
+			blake3_hasher_finalize(&hasher, reinterpret_cast<uint8_t*>(&state->outHash), sizeof(CState::hash_t)); // finalize output hash for whole image given all hashes
+		}
+
+		return true;
+	}
+	static inline bool execute(state_type* state)
+	{
+		return execute(core::execution::seq, state);
+	}
+
+private:
+
+	struct ScratchMap
+	{
+		asset::SBufferRange<asset::ICPUBuffer> hashes; // hashes, single hash is obtained from given miplevel & layer, full hash for an image is a hash of this hash buffer
+		asset::SBufferRange<asset::ICPUBuffer> hashers; // hashers, used to produce a hash
+		asset::SBufferRange<asset::ICPUBuffer> flatten; // tightly packed texels from input, no memory gaps
```

Comment:
its weird to use `ICPUBuffer` which is an asset, to represent what is essentially a span<uint8_t>

Reply:
actually a `std::span<hash_t>`, `std::span<blake3_hasher>` , the last one can stay an ICPUBuffer